src/stdio.jl

```diff
@@ -27,36 +27,51 @@ Base.setup_stdio(io::IJuliaStdio, readable::Bool) = Base.setup_stdio(io.io.io, r
 
 for s in ("stdout", "stderr", "stdin")
     f = Symbol("redirect_", s)
-    S = QuoteNode(Symbol(uppercase(s)))
+    Sq = QuoteNode(Symbol(uppercase(s)))
+    sq = QuoteNode(Symbol(s))
     @eval function Base.$f(io::IJuliaStdio)
         io[:jupyter_stream] != $s && throw(ArgumentError(string("expecting ", $s, " stream")))
-        eval(Base, Expr(:(=), $S, io))
+        if isdefined(Base, :stdout)
+            # On Julia 0.7, we need to override the global logger as well
+            logger = Base.CoreLogging._global_logstate.logger
+
+            # Override logging if it's pointing at the default stderr
+            if logger.stream == Base.stderr
```

Comment:
Actually, there is something bugged here; this is only true on the first call, don't we need to reset it somewhere @staticfloat ? On this branch; with a `@show logger.stream == Base.stderr` inserted:
```
julia> using IJulia

julia> mktemp() do path, io
           redirect_stderr(IJulia.IJuliaStdio(io, "stderr")) do
               @warn("warn")
           end
           flush(io)
           seek(io, 0)
           @show read(io, String)
       end
logger.stream == Base.stderr = true
read(io, String) = "\e[33m\e[1m┌ \e[22m\e[39m\e[33m\e[1mWarning: \e[22m\e[39mwarn\n\e[33m\e[1m└ \e[22m\e[39m\e[90m@ Main REPL[2]:3\e[39m\n"
"\e[33m\e[1m┌ \e[22m\e[39m\e[33m\e[1mWarning: \e[22m\e[39mwarn\n\e[33m\e[1m└ \e[22m\e[39m\e[90m@ Main REPL[2]:3\e[39m\n"

julia> mktemp() do path, io
           redirect_stderr(IJulia.IJuliaStdio(io, "stderr")) do
               @warn("warn")
           end
           flush(io)
           seek(io, 0)
           @show read(io, String)
       end
logger.stream == Base.stderr = false
read(io, String) = ""
```

Reply:
Perhaps it does not matter; this seems to be called only once per session anyways and it works in a notebook
![warning](https://user-images.githubusercontent.com/11698744/42364783-b3a1be48-80fc-11e8-98d2-2cc4e9939e09.png)

Reply:
Would it not be simpler to just replace the global logger instead of checking the internal state?